File: Data Structures & Algorithms/trapping-rain-water/submission-2.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int s = 0;
        if (n == 0)
            return 0;
        int l = 0, r = n-1;
        int lh = height[l], rh = height[r];
        while (l < r) {
            if (lh < rh) {
                                l++;

                lh = max(lh, height[l]);
                s+= lh - height[l];
            } else {
                                r--;

                rh = max(rh, height[r]);
                s+= rh - height[r];
            }

        }
        return s;
    }
};

File: Data Structures & Algorithms/network-delay-time/submission-2.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        adj = defaultdict(list)
        for u, v, w in times:
            adj[u].append((v,w))
        dist = {node: float('inf') for node in range(1,n+1)}

        def dfs(node, t):
            if t >= dist[node]:
                return
            dist[node] = t
            for v, w in adj[node]:
                dfs(v, w+t)
        
        dfs(k,0)
        res = max(dist.values())
        return res if res < float('inf') else -1
        